### PR TITLE
feat: added custom grpc resolver

### LINKED ISSUE #995: [flagd] Support custom schemes (name resolvers)
As described [here](https://github.com/open-feature/flagd/blob/main/docs/reference/specifications/proposal/rfc-grpc-custom-name-resolver.md), we wish to support [custom name revolvers](https://grpc.io/docs/guides/custom-name-resolution/) for additional control and configuration according to user's infrastructure. In order to make use of this, we must support the creation of channels created with [NettyChannelBuilder.forTarget](https://grpc.github.io/grpc-java/javadoc/io/grpc/netty/NettyChannelBuilder.html#forTarget(java.lang.String)), and expose an option in the flagd that use a "target string" instead of a host/port combination.

Requirements:

- optional usage of channel built using `.forTarget` (see: https://github.com/open-feature/java-sdk-contrib/blob/4197d3231d68fb686b61ac1ca5eb8c278044a0e3/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java#L30)
- changes in options to allow specifying a target string
  - support env-var based configuration as well, as with existing options
  - consistent with https://github.com/open-feature/go-sdk-contrib/issues/585
- no breaking changes
- associated unit tests (e2e not necessary)

cc @pradeepbbl

### PR DESCRIPTION
**Pull request by pradeepbbl**
## This PR

Added custom gRPC resolver to support envoy proxy

- support gRPC custom resolver using new config options `tragetUri`   

### Related Issues
Fixes #995 

### How to test
Unit test are already added for the custom resolver for integration test you need a working `envoy` proxy support or any of the supported core resolver mentioned [here](https://grpc.io/docs/guides/custom-name-resolution/#overview) 

```java (in-process builder)
FlagdOptions.builder()
      .targetUri("envoy://localhost:9211/b-features-api.service")
      .selector("test-project")
      .resolverType(Config.Resolver.IN_PROCESS)
      .build();
```

```log
DEBUG [2024-10-08 00:00:37,661] io.grpc.netty.NettyClientHandler: [id: 0xef9bc3d3, L:/127.0.0.1:35368 - R:localhost/127.0.0.1:9211] OUTBOUND HEADERS: streamId=3 headers=GrpcHttp2OutboundHeaders[:authority: b-features-api.service, :path: /flagd.sync.v1.FlagSyncService/SyncFlags, :method: POST, :scheme: http, content-type: application/grpc, te: trailers, user-agent: grpc-java-netty/1.58.0, grpc-accept-encoding: gzip, grpc-timeout: 599837m] streamDependency=0 weight=16 exclusive=false padding=0 endStream=false
DEBUG [2024-10-08 00:00:37,664] io.grpc.netty.NettyClientHandler: [id: 0xef9bc3d3, L:/127.0.0.1:35368 - R:localhost/127.0.0.1:9211] OUTBOUND HEADERS: streamId=5 headers=GrpcHttp2OutboundHeaders[:authority: b-features-api.service, :path: /flagd.sync.v1.FlagSyncService/GetMetadata, :method: POST, :scheme: http, content-type: application/grpc, te: trailers, user-agent: grpc-java-netty/1.58.0, grpc-accept-encoding: gzip, grpc-timeout: 9969998u] streamDependency=0 weight=16 exclusive=false padding=0 endStream=false
DEBUG [2024-10-08 00:00:37,666] io.grpc.netty.NettyClientHandler: [id: 0xef9bc3d3, L:/127.0.0.1:35368 - R:localhost/127.0.0.1:9211] OUTBOUND DATA: streamId=5 padding=0 endStream=true length=5 bytes=0000000000
DEBUG [2024-10-08 00:00:37,667] io.grpc.netty.NettyClientHandler: [id: 0xef9bc3d3, L:/127.0.0.1:35368 - R:localhost/127.0.0.1:9211] OUTBOUND DATA: streamId=3 padding=0 endStream=true length=15 bytes=000000000a12086665617475726573
DEBUG [2024-10-08 00:00:37,719] io.grpc.netty.NettyClientHandler: [id: 0xef9bc3d3, L:/127.0.0.1:35368 - R:localhost/127.0.0.1:9211] INBOUND HEADERS: streamId=5 headers=GrpcHttp2ResponseHeaders[:status: 200, server: envoy, date: Tue, 08 Oct 2024 00:00:37 GMT, content-type: application/grpc, content-length: 0, grpc-status: 12, grpc-message: Method flagd.sync.v1.FlagSyncService/GetMetadata is unimplemented, x-envoy-upstream-service-time: 43] padding=0 endStream=true
DEBUG [2024-10-08 00:00:37,722] io.grpc.netty.NettyClientHandler: [id: 0xef9bc3d3, L:/127.0.0.1:35368 - R:localhost/127.0.0.1:9211] INBOUND HEADERS: streamId=3 headers=GrpcHttp2ResponseHeaders[:status: 200, server: envoy, date: Tue, 08 Oct 2024 00:00:37 GMT, content-type: application/grpc, grpc-encoding: gzip, grpc-accept-encoding: gzip, x-envoy-upstream-service-time: 44] padding=0 endStream=false

```

### CODE REVIEW

File: providers/flagd/README.md [hunk 1]
```diff
@@ -239,3 +241,17 @@ FlagdProvider flagdProvider = new FlagdProvider(options);
 Please refer [OpenTelemetry example](https://opentelemetry.io/docs/instrumentation/java/manual/#example) for best practice guidelines.
 
 Provider telemetry combined with [flagd OpenTelemetry](https://flagd.dev/reference/monitoring/#opentelemetry) allows you to have distributed traces.
+
+### TragetUri Support (gRPC name resolution)
```
Comment (beeme1mr) [inline providers/flagd/README.md]:
```suggestion
### Target URI Support (gRPC name resolution)
```

File: providers/flagd/README.md [hunk 2]
```diff
@@ -239,3 +241,17 @@ FlagdProvider flagdProvider = new FlagdProvider(options);
...
+
+The ``tagetUri`` is meant for gRPC custom name resolution (default is `dns`), this allows users to use different
```
Comment (beeme1mr) [inline providers/flagd/README.md]:
```suggestion
The `targetUri` is meant for gRPC custom name resolution (default is `dns`), this allows users to use different
```

File: providers/flagd/README.md [hunk 3]
```diff
@@ -239,3 +241,17 @@ FlagdProvider flagdProvider = new FlagdProvider(options);
...
+resolution method e.g. `xds`. Currently, we are supporting all [core resolver](https://grpc.io/docs/guides/custom-name-resolution/)
+and one custom resolver for ``envoy`` proxy resolution. For more details please refer the
```
Comment (beeme1mr) [inline providers/flagd/README.md]:
```suggestion
and one custom resolver for `envoy` proxy resolution. For more details, please refer the 
```

**Review by beeme1mr — APPROVED** [review]

**Comment by toddbaert** [comment]
Hey @pradeepbbl ! Thanks a lot for this. I was hoping to get to it this week, but I won't be able to. I will review next week. Sorry about that.

**Comment by pradeepbbl** [comment]
> Hey @pradeepbbl ! Thanks a lot for this. I was hoping to get to it this week, but I won't be able to. I will review next week. Sorry about that.

No worries take your time meanwhile I'm exploring way to have e2e test using local envoy container

**Comment by pradeepbbl** [comment]
added a new e2e test coverage for `envoy` custom resolver

Test Name: **RunFlagdInProcessEnvoyCucumberTest**

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java [hunk 4]
```diff
@@ -50,9 +54,21 @@ public static ManagedChannel nettyChannel(final FlagdOptions options) {
 
         // build a TCP socket
         try {
+            // Register custom resolver
+            if (isEnvoyTarget(options.getTargetUri())) {
+                NameResolverRegistry.getDefaultRegistry().register(new EnvoyResolverProvider());
```
Comment (toddbaert) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java:59]: Good idea to only register the resolver if the option is set.

File: providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/ContainerConfig.java [hunk 5]
```diff
@@ -58,6 +69,22 @@ public static GenericContainer flagd(boolean unstable) {
...
     }
 
+
+    /**
+     * @return a {@link org.testcontainers.containers.GenericContainer} instance of envoy container using
+     * flagd sync service as backend expose on port 9211
+     *
+     */
+    public static GenericContainer envoy() {
+        final String container = "envoyproxy/envoy:v1.31.0";
```
Comment (toddbaert) [inline providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/ContainerConfig.java:79]: @pradeepbbl I have to say I'm quite impressed that you went as far as creating e2e tests for this. Excellent work, very cool.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java [hunk 6]
```diff
@@ -78,6 +94,46 @@ public static ManagedChannel nettyChannel(final FlagdOptions options) {
...
+    }
+
+    private static  boolean isValidTargetUri(String targetUri) {
+        if (targetUri == null) {
+            return false;
+        }
+
+        try {
+            final String scheme = new URI(targetUri).getScheme();
+            if (scheme.equals("envoy") || scheme.equals("dns") || scheme.equals("xds") || scheme.equals("uds")) {
```
Comment (aepfli) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java]: [suggestion] should we maybe put those strings into an enum, or maybe there is an existing enum - i am not a big fan of strings, and specially if they are duplicated.

**Review by aepfli — APPROVED** [review]
nice, love the e2e tests, just added a suggestion regarding enums.

**Comment by toddbaert** [comment]
I only see one small issue: the way `GLUE_PROPERTY_NAME` name works, for example [here](https://github.com/open-feature/java-sdk-contrib/pull/1008/files#diff-1151aff13c6105b60e85b16a2b7dc29ebf06b24bae2d4f75f348dde26a5f208aR23), it resolves all subclasses under that package: so your `RunFlagdInProcessEnvoyCucumberTest` class is running BOTH the `dev.openfeature.contrib.providers.flagd.e2e.process.envoy.FlagdInProcessSetup` AND the `dev.openfeature.contrib.providers.flagd.e2e.process.FlagdInProcessSetup`. To fix this, just make these classes/packages siblings (and probably rename your setup class to something like `FlagdInProcessEnvoySetup` or something like that).

**Review by toddbaert — CHANGES_REQUESTED** [review]
Excellent work, thanks so much. [This](https://github.com/open-feature/java-sdk-contrib/pull/1008#issuecomment-2414600783) is my only blocking change (which is small and I would do myself, but I can't edit your branch)

File: providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/RunFlagdInProcessEnvoyCucumberTest.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import org.junit.platform.suite.api.IncludeEngines;
+import org.junit.platform.suite.api.SelectClasspathResource;
+import org.junit.platform.suite.api.Suite;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import static io.cucumber.junit.platform.engine.Constants.GLUE_PROPERTY_NAME;
+import static io.cucumber.junit.platform.engine.Constants.PLUGIN_PROPERTY_NAME;
+
+/**
+ * Class for running the tests associated with "stable" e2e tests (no fake disconnection) for the in-process provider
```
Comment (toddbaert) [inline providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/RunFlagdInProcessEnvoyCucumberTest.java:14]:
```suggestion
 * Class for running the tests associated with "stable" e2e tests (no fake disconnection) for the in-process provider, and an intervening envoy proxy
```

**Comment by toddbaert** [comment]
@pradeepbbl I merged another PR that results in some (I believe minor) conflicts here, mostly with the setup of the providers for e2e tests. Just be sure to use:

```java
.deadline(1000)
.streamDeadlineMs(0) // this makes reconnect tests more predictable
```

for the options in those.

I _think_ that's the only thing to note, but if you have issues with the conflicts or e2e tests not working, please let me know and I will do what I can to help.

File: providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/ContainerConfig.java (hunk 5 above)
Comment (pradeepbbl) [reply providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/ContainerConfig.java:79]: thank you

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java (hunk 6 above)
Comment (pradeepbbl) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java]: ack

Comment (pradeepbbl) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java]: @aepfli thanks for the suggestion .. added a new enum class `SupportedScheme` and replace the string with enum

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java (hunk 4 above)
Comment (pradeepbbl) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java:59]: thanks my idea was not to touch the core resolver when it's not needed

**Comment by pradeepbbl** [comment]
> I only see one small issue: the way `GLUE_PROPERTY_NAME` name works, for example [here](https://github.com/open-feature/java-sdk-contrib/pull/1008/files#diff-1151aff13c6105b60e85b16a2b7dc29ebf06b24bae2d4f75f348dde26a5f208aR23), it resolves all subclasses under that package: so your `RunFlagdInProcessEnvoyCucumberTest` class is running BOTH the `dev.openfeature.contrib.providers.flagd.e2e.process.envoy.FlagdInProcessSetup` AND the `dev.openfeature.contrib.providers.flagd.e2e.process.FlagdInProcessSetup`. To fix this, just make these classes/packages siblings (and probably rename your setup class to something like `FlagdInProcessEnvoySetup` or something like that).

@toddbaert thanks as suggested fixed the `GLUE_PROPERTY_NAME` and also the merge conflict, please have a look when you have free time.

**Comment by pradeepbbl** [comment]
> Excellent work, thanks so much. [This](https://github.com/open-feature/java-sdk-contrib/pull/1008#issuecomment-2414600783) is my only blocking change (which is small and I would do myself, but I can't edit your branch)

yeah sorry for that as per the company policy all our open-source contribution need to be done through company github org

**Review by aepfli — APPROVED** [review]

**Review by toddbaert — APPROVED** [review]
Nothing else to say but amazing work!

I'll merge this today unless anyone objects, and then release.

**Comment by toddbaert** [comment]
https://github.com/open-feature/java-sdk-contrib/pull/1022#issuecomment-2417567334

**Pull request merged**